shell/platform/android/test/io/flutter/embedding/android/FlutterViewTest.java

```diff
@@ -649,6 +651,69 @@ public void systemInsetDisplayCutoutSimple() {
     assertEquals(100, viewportMetricsCaptor.getValue().viewInsetTop);
   }
```

Comment:
Based on the issue https://github.com/flutter/flutter/issues/155658 I think there are more regression tests we should include. 
From my reading this test is verifying that cutouts when triggered how we expect show up in viewpoint metrics. 

After Reading the bug i am thinking we should have a rotation test that verifies that cutouts are not "left" after the hardware has changed orientation. Additionally there is logic in this pr to combine displayCutouts and displayFeatures, I would expect a test to modify one and assert the other was preserved and vice versa *and* a test to show that if both are modfied the joint list is updated correctly. This second (of 3) change might be possible without adding a new test but by adding asserts to existing tests or it might require a new test depending on the existing coverage.

Reply:
For testing the rotation as you mention, this sounds like it would require an integration test to run on an emulator; I do not think we can unit-test it, as changing the orientation in a unit test will have no effect on the value returned by the stubbed methods like we are using here. Do you agree with that? And if so, are there existing integration tests in this repo that perform anything similar?

Reply:
I am hoping we can not do an integration test but instead do a roboletric test that runs on an activity. Possibly by modifying the orientation then recreating the activity. 
See https://stackoverflow.com/a/26468296 for inspiration.

Reply:
We can change the orientation in a robolectric unit test, but I still fail to see how that will help us for a unit test where the display cutouts are provided through a mock with stubbed methods for returning its metrics. The closest thing I can find is that maybe we can use robolectric's ShadowDisplay (which is deprecated and planned for removal), and to access the ShadowDisplay in the mock's stubs. I have not yet tested this approach, but if that would actually work, I am still unsure if it even reflects the behavior we want to test within reason.

Reply:
I go back and forth on if an integration test is appropriate here in general we get more flakes with an integration test but also this is something our users are expecting us to handle. 

I think your suggestion to use shadowDisplay is interesting but if it is depercrated then not worth the time.

Reply:
Even disregarding the deprecation, I don't think using shadow displays would actually test what we are looking to test. The display cutouts are already repositioned wrt orientation by the system by the time the containing insets are passed to `onApplyWindowInsets`. All we would really accomplish by using the shadow display would be to emulate some of that logic ourselves.

Reply:
Let go ahead and do a integration test that validates rotation and cutout. To me it is right on the edge of what needs /justifies an integration test but rotation logic is a classic cause of android errors and I think it is likely that if we have another rotation regression around display logic we will be thankful the test already exists.

Reply:
@reidbaker By this, do you mean a scenario in the scenario app? And if so, how do we specify/enforce device features such as cutouts when the app is run? I have been looking through the scenario app code and CI files for the past couple days and still can't identify what provides the "physical" characteristics of the device that runs the tests.

Reply:
I think the rotation test might need to be in the framework but as to how to define physical devices I am not aware of an api that lets you specific physical traits. There are developer settings that might be able to be set via adb that can simulate notches also we have access to physical devices and I think you can force a device to rotate also via adb. 

Now that you are digging into it I think there is additional value to the test if only for the reason of talking about how we can test things like rotation.
It could be recency bias but I think the native_driver_tests that @matanlurey added could be expanded to include rotation.

Reply:
There are [rotation tests](https://github.com/flutter/flutter/blob/1cc8a07ace699de12fcd31ac9e8fcfa14f1017e2/dev/integration_tests/native_driver_test/test_driver/platform_view_blue_orange_gradient_main_test.dart#L44-L61) if that's what you mean.

Reply:
@reidbaker do you know if we have any existing examples of forcing a test to run on a specific physical device and/or requiring a test to be run only in the context of certain adb commands?

Reply:
Also I want to make sure I document here before I forget that you can get a list of cutout overlays with

    adb shell cmd overlay list | grep "display\.cutout"

and then dis/enable one with

    adb shell cmd overlay <enable/disable> <cutout package name>

Reply:
@reidbaker Actually, I have an integration test on a local branch that can test this behavior. However, as the current state of this behavior is bugged, my understanding is that this PR will need to land first with only the engine-level unit tests before we can add this integration test to the framework. Does this sound right? If so, do you have any more pending thoughts or concerns for this PR?

Reply:
Let me rephrase to see if I understand. You have an integration test on a branch of flutter/flutter that fails with the current behavior and passes with a custom engine running this pr. If that is the case then yes we can land this pr then your other pr. That said, let's put that second pr up for review so that reviewers can look at what is covered in each. 

TBH having the test fail against flutter/flutter and pass with this patch is exactly the type of test we want here thank you for writing that test.